RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c

```diff
@@ -0,0 +1,1336 @@
+
+**/
+EFI_STATUS
+EFIAPI
+RedfishHttpEntryPoint (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+  VOID        *Registration;
+
+  if (mRedfishHttpCachePrivate != NULL) {
+    return EFI_ALREADY_STARTED;
+  }
+
+  mRedfishHttpCachePrivate = AllocateZeroPool (sizeof (REDFISH_HTTP_CACHE_PRIVATE));
+  if (mRedfishHttpCachePrivate == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  //
+  // Initial cache list and protocol instance.
+  //
+  mRedfishHttpCachePrivate->Signature   = REDFISH_HTTP_DRIVER_SIGNATURE;
+  mRedfishHttpCachePrivate->ImageHandle = ImageHandle;
+  CopyMem (&mRedfishHttpCachePrivate->Protocol, &mEdkIIRedfishHttpProtocol, sizeof (EDKII_REDFISH_HTTP_PROTOCOL));
+  mRedfishHttpCachePrivate->CacheList.Capacity = REDFISH_HTTP_CACHE_LIST_SIZE;
+  mRedfishHttpCachePrivate->CacheList.Count    = 0x00;
+  mRedfishHttpCachePrivate->CacheDisabled      = PcdGetBool (PcdHttpCacheDisabled);
+  InitializeListHead (&mRedfishHttpCachePrivate->CacheList.Head);
+
+  //
+  // Get retry settings
+  //
+  mRedfishHttpCachePrivate->RetrySetting.MaximumRetryGet    = PcdGet16 (PcdHttpGetRetry);
+  mRedfishHttpCachePrivate->RetrySetting.MaximumRetryPut    = PcdGet16 (PcdHttpPutRetry);
+  mRedfishHttpCachePrivate->RetrySetting.MaximumRetryPatch  = PcdGet16 (PcdHttpPatchRetry);
+  mRedfishHttpCachePrivate->RetrySetting.MaximumRetryPost   = PcdGet16 (PcdHttpPostRetry);
+  mRedfishHttpCachePrivate->RetrySetting.MaximumRetryDelete = PcdGet16 (PcdHttpDeleteRetry);
+  mRedfishHttpCachePrivate->RetrySetting.RetryWait          = PcdGet16 (PcdHttpRetryWaitInSecond) * 1000000U;
+
+  //
+  // Install the gEdkIIRedfishHttpProtocolGuid onto Handle.
+  //
+  Status = gBS->InstallMultipleProtocolInterfaces (
+                  &mRedfishHttpCachePrivate->ImageHandle,
+                  &gEdkIIRedfishHttpProtocolGuid,
+                  &mRedfishHttpCachePrivate->Protocol,
+                  NULL
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: cannot install Redfish http protocol: %r\n", __func__, Status));
+    RedfishHttpDriverUnload (ImageHandle);
+    return Status;
+  }
+
+  //
+  // Install protocol notification if credential protocol is installed.
+  //
+  mRedfishHttpCachePrivate->NotifyEvent = EfiCreateProtocolNotifyEvent (
+                                            &gEdkIIRedfishCredentialProtocolGuid,
+                                            TPL_CALLBACK,
+                                            CredentialProtocolInstalled,
+                                            mRedfishHttpCachePrivate,
+                                            &Registration
+                                            );
+  if (mRedfishHttpCachePrivate->NotifyEvent == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: failed to create protocol notification for gEdkIIRedfishCredentialProtocolGuid\n", __func__));
+    ASSERT (FALSE);

```

Comment:
we should also return EFI error in the case ASSERT is implicitly disabled.  Also free mRedfishHttpCachePrivate memory.